### PR TITLE
Check the member UUIDs that the schema is replicated in the client schema service [API-1646]

### PR DESCRIPTION
**Pull request by harunalpak**
Related java issue: https://github.com/hazelcast/hazelcast/pull/22197

One of the test failed after code changes. I skipped this test. I will find a solution for test.

### CODE REVIEW

File: src/HazelcastClient.ts [hunk 1]
```diff
@@ -146,7 +146,12 @@ export class HazelcastClient {
         this.loggingService = new LoggingService(this.config.customLogger,
             this.config.properties['hazelcast.logging.level'] as string);
         const logger = this.loggingService.getLogger();
-        this.schemaService = new SchemaService(() => this.invocationService, logger);
+        this.schemaService = new SchemaService(
```
Comment (srknzl) [inline src/HazelcastClient.ts:149]: You can move clusterService and clusterFailoverService construction above and use clusterService directly: 

```diff
diff --git a/src/HazelcastClient.ts b/src/HazelcastClient.ts
index a39d6eb5..9d35a447 100644
--- a/src/HazelcastClient.ts
+++ b/src/HazelcastClient.ts
@@ -146,10 +146,16 @@ export class HazelcastClient {
         this.loggingService = new LoggingService(this.config.customLogger,
             this.config.properties['hazelcast.logging.level'] as string);
         const logger = this.loggingService.getLogger();
+        this.clusterFailoverService = this.initClusterFailoverService();
+        this.clusterService = new ClusterService(
+            this.config,
+            logger,
+            this.clusterFailoverService
+        );
         this.schemaService = new SchemaService(
-            this.config, 
-            () => this.clusterService, 
-            () => this.invocationService, 
+            this.config,
+            this.clusterService,
+            () => this.invocationService,
             logger
         );
         this.serializationService = new SerializationServiceV1(this.config.serialization, this.schemaService);
@@ -166,12 +172,6 @@ export class HazelcastClient {
             this.config.lifecycleListeners,
             logger
         );
-        this.clusterFailoverService = this.initClusterFailoverService();
-        this.clusterService = new ClusterService(
-            this.config,
-            logger,
-            this.clusterFailoverService
-        );
         this.connectionRegistry = new ConnectionRegistryImpl(
             this.config.connectionStrategy.asyncStart,
             this.config.connectionStrategy.reconnectMode,

```

File: src/config/Config.ts [hunk 2]
```diff
@@ -178,6 +178,7 @@ const DEFAULT_PROPERTIES: Properties = {
     'hazelcast.client.heartbeat.interval': 5000,
     'hazelcast.client.heartbeat.timeout': 60000,
     'hazelcast.client.invocation.retry.pause.millis': 1000,
+    'hazelcast.client.schema.max.put.retry.count': 100,
```
Comment (srknzl) [inline src/config/Config.ts:181]: Can you update ConfigBuilder.validateProperty and ConfigBuilderTest as well? We should add validation and test for validation. 

@mdumandag Should we document this property? Is it public?

File: src/proxy/Ringbuffer.ts [hunk 3]
```diff
@@ -156,13 +156,22 @@ export interface Ringbuffer<E> extends DistributedObject {
...
      * @param minCount minimum number of items to be read.
      * @param maxCount maximum number of items to be read.
+     * @param filter optional filter to be applied to the items
      * @throws RangeError if `startSequence` is smaller than `0`,
      *                      or if `minCount` smaller than `0`,
      *                      or if `minCount` larger than `maxCount`,
      *                      or if `maxCount` larger than `1000` (to prevent overloading)
      */
-    readMany(startSequence: number | Long, minCount: number, maxCount: number): Promise<ReadResultSet<E>>;
+    readMany(startSequence: number | Long, minCount: number, maxCount: number, filter?: any): Promise<ReadResultSet<E>>;
```
Comment (srknzl) [inline src/proxy/Ringbuffer.ts:176]: Is this change related?

File: test/integration/backward_compatible/parallel/ringbuffer/RingbufferProxyTest.js [hunk 4]
```diff
@@ -158,4 +158,13 @@ describe('RingbufferProxyTest', function () {
         const size = await rb.size();
         expect(size.toNumber()).to.equal(2);
     });
+
+    it('readMany correctly works with a filter', async function () {
```
Comment (srknzl) [inline test/integration/backward_compatible/parallel/ringbuffer/RingbufferProxyTest.js:162]: Is this change related?

File: src/serialization/compact/SchemaService.ts [hunk 5]
```diff
@@ -106,4 +129,39 @@ export class SchemaService {
...
         return this.getInvocationService().invokeUrgent(invocation).then(() => {});
     }
+
+    async replicateSchemaInCluster(schema: Schema): Promise<boolean> {
+        let clientMessage = ClientSendSchemaCodec.encodeRequest(schema);
+        outer:
+        for (let index = 0; index < this.maxPutRetryCount; index++) {
+            const invocation = new Invocation(this.getInvocationService(), clientMessage);
+            const response = await this.getInvocationService().invoke(invocation);
+            const replicatedMemberUuids = UuidUtil.convertUUIDSetToStringSet(ClientSendSchemaCodec.decodeResponse(response));
```
Comment (srknzl) [inline src/serialization/compact/SchemaService.ts:139]: Why do we convert UUID set to a string set?

File: src/serialization/compact/SchemaService.ts [hunk 6]
```diff
@@ -67,18 +82,26 @@ export class SchemaService {
...
+                `The schema ${schema} cannot be replicated in the cluster, after ${this.maxPutRetryCount}  retries. 
+                It might be the case that the client is connected to the two halves of the cluster that 
+                is experiencing a split-brain, and continue putting the data associated with that schema might 
+                result in data loss. It might be possible to replicate the schema after some time, when 
+                the cluster is healed.`
+            );
+        } else {
             this.putIfAbsent(schema);
-        });
+        }
```
Comment (srknzl) [inline src/serialization/compact/SchemaService.ts:104]:
```suggestion
        }
```

File: src/serialization/compact/SchemaService.ts [hunk 7]
```diff
@@ -67,18 +82,26 @@ export class SchemaService {
     /**
      * Puts the schema with the given id to the cluster.
      */
-    put(schema: Schema): Promise<void> {
+    async put(schema: Schema): Promise<void> {
```
Comment (srknzl) [inline src/serialization/compact/SchemaService.ts:85]: Please use then/catch as async/await has more overhead.

File: src/serialization/compact/SchemaService.ts [hunk 8]
```diff
@@ -67,18 +82,26 @@ export class SchemaService {
...
             return Promise.resolve();
         }
-        const message = ClientSendSchemaCodec.encodeRequest(schema);
-        const invocation = new Invocation(this.getInvocationService(), message);
-        return this.getInvocationService().invoke(invocation).then(() => {
+
+        const result = await this.replicateSchemaInCluster(schema);
+        if (!result) {
+            throw new IllegalStateError(
+                `The schema ${schema} cannot be replicated in the cluster, after ${this.maxPutRetryCount}  retries.
```
Comment (srknzl) [inline src/serialization/compact/SchemaService.ts:96]: schema.toString does not exist, and this won't look good. What about "the schema with typename schema.typeName"?

File: src/serialization/compact/SchemaService.ts [hunk 9]
```diff
@@ -106,4 +129,39 @@ export class SchemaService {
         const invocation = new Invocation(this.getInvocationService(), message);
         return this.getInvocationService().invokeUrgent(invocation).then(() => {});
     }
+
+    async replicateSchemaInCluster(schema: Schema): Promise<boolean> {
```
Comment (srknzl) [inline src/serialization/compact/SchemaService.ts:133]: Without async/await please

**Review by srknzl — CHANGES_REQUESTED** [review]

File: src/HazelcastClient.ts (hunk 1 above)
Comment (harunalpak) [reply src/HazelcastClient.ts:149]: Fixed

File: test/integration/backward_compatible/parallel/ringbuffer/RingbufferProxyTest.js (hunk 4 above)
Comment (harunalpak) [reply test/integration/backward_compatible/parallel/ringbuffer/RingbufferProxyTest.js:162]: Not related. I dont know, how it was included in this PR? I will fix it.

File: src/config/Config.ts (hunk 2 above)
Comment (harunalpak) [reply src/config/Config.ts:181]: Fixed

File: src/proxy/Ringbuffer.ts (hunk 3 above)
Comment (harunalpak) [reply src/proxy/Ringbuffer.ts:176]: Not related. I will fix it.

File: src/serialization/compact/SchemaService.ts (hunk 5 above)
Comment (harunalpak) [reply src/serialization/compact/SchemaService.ts:139]: Because I could not use `Set.has` for objects.  And `ClientSendSchemaCodec.decodeResponse` return `Set<UUID>`.
Because of that I convert from it to `Set<string>`.

Comment (srknzl) [reply src/serialization/compact/SchemaService.ts:138]: You can use objects. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/has#using_the_has_method

File: src/serialization/compact/SchemaService.ts (hunk 8 above)
Comment (harunalpak) [reply src/serialization/compact/SchemaService.ts:95]: Fixed

File: src/serialization/compact/SchemaService.ts (hunk 9 above)
Comment (harunalpak) [reply src/serialization/compact/SchemaService.ts:132]: Fixed

File: src/serialization/compact/SchemaService.ts (hunk 7 above)
Comment (harunalpak) [reply src/serialization/compact/SchemaService.ts:84]: Fixed

File: src/serialization/compact/SchemaService.ts (hunk 5 above)
Comment (srknzl) [reply src/serialization/compact/SchemaService.ts]: Turned out that we can't because SameValue semantics is used and the same object should be passed to the has() method, which is not possible

File: src/serialization/compact/SchemaService.ts [hunk 10]
```diff
@@ -74,11 +89,19 @@ export class SchemaService {
...
+                    `The schema ${schema.typeName} cannot be replicated in the cluster, after ${this.maxPutRetryCount}  retries. 
+                    It might be the case that the client is connected to the two halves of the cluster that 
+                    is experiencing a split-brain, and continue putting the data associated with that schema might 
+                    result in data loss. It might be possible to replicate the schema after some time, when 
+                    the cluster is healed.`
+                );
+            } else {
+                this.putIfAbsent(schema);
+            }
+        }).catch((err) => { console.error(err); });
```
Comment (srknzl) [inline src/serialization/compact/SchemaService.ts:104]: We don't use console.error directly. We can either ignore the error completely, by `catch(() => {})` or if we are sure that the error handling of the promise returned by this method will be handled, we can skip putting a catch(). We can also log the error if we want via the logger.  

Right now, all the register errors are propagated to the user via async APIs. If the user doesn't register an error handler the program may crash if the user is using a new Node.js runtime. I think we can remove the catch block for this promise since we cannot continue without a schema. User should get an error.

File: src/serialization/compact/SchemaService.ts [hunk 11]
```diff
@@ -106,4 +129,42 @@ export class SchemaService {
         const invocation = new Invocation(this.getInvocationService(), message);
         return this.getInvocationService().invokeUrgent(invocation).then(() => {});
     }
+
+    replicateSchemaInCluster(schema: Schema): Promise<boolean> {
+        const clientMessage = ClientSendSchemaCodec.encodeRequest(schema);
+        return this.retryMaxPutRetryCount(clientMessage, 0);
+    }
+
+    private retryMaxPutRetryCount(clientMessage: ClientMessage, index: number) : Promise<boolean> {
```
Comment (srknzl) [inline src/serialization/compact/SchemaService.ts:138]:
```suggestion
    private retryMaxPutRetryCount(clientMessage: ClientMessage, currentRetryCount: number) : Promise<boolean> {
```

or it can be `retryCount`. IMO, index is too general

File: src/serialization/compact/SchemaService.ts [hunk 12]
```diff
@@ -106,4 +129,42 @@ export class SchemaService {
...
+                const replicatedMemberUuids = UuidUtil.convertUUIDSetToStringSet(ClientSendSchemaCodec.decodeResponse(response));
+                const members = this.getClusterService().getMembers();
+                for (const member of members) {
+                    if (!replicatedMemberUuids.has(member.uuid.toString())) {
+                        // There is a member in our member list that the schema
+                        // is not known to be replicated yet. We should retry
+                        // sending it in a random member.
+    
+                        return delayedPromise(this.retryPauseMillis)
+                        .then(() => {
```
Comment (srknzl) [inline src/serialization/compact/SchemaService.ts:154]:
```suggestion
                        return delayedPromise(this.retryPauseMillis).then(() => {
```

File: src/serialization/compact/SchemaService.ts [hunk 13]
```diff
@@ -106,4 +129,42 @@ export class SchemaService {
...
+
+    replicateSchemaInCluster(schema: Schema): Promise<boolean> {
+        const clientMessage = ClientSendSchemaCodec.encodeRequest(schema);
+        return this.retryMaxPutRetryCount(clientMessage, 0);
+    }
+
+    private retryMaxPutRetryCount(clientMessage: ClientMessage, index: number) : Promise<boolean> {
+        if (index === this.maxPutRetryCount) {
+            return Promise.resolve(false);
+        } else {
```
Comment (srknzl) [inline src/serialization/compact/SchemaService.ts:141]: we can remove the else block here, not to indent the long code below.

File: src/serialization/compact/SchemaService.ts [hunk 14]
```diff
@@ -74,11 +89,19 @@ export class SchemaService {
...
+                throw new IllegalStateError(
+                    `The schema ${schema.typeName} cannot be replicated in the cluster, after ${this.maxPutRetryCount}  retries. 
+                    It might be the case that the client is connected to the two halves of the cluster that 
+                    is experiencing a split-brain, and continue putting the data associated with that schema might 
+                    result in data loss. It might be possible to replicate the schema after some time, when 
+                    the cluster is healed.`
+                );
+            } else {
+                this.putIfAbsent(schema);
+            }
```
Comment (srknzl) [inline src/serialization/compact/SchemaService.ts:103]:
```suggestion
            }
            this.putIfAbsent(schema);
```

**Review by srknzl — CHANGES_REQUESTED** [review]

**Comment by srknzl** [comment]
@harunalpak Can you check if we can increase coverage easily by adding some tests? There are warnings regarding coverage

**Comment by srknzl** [comment]
The coverage drop is 0.11%, which is not acceptable

File: src/serialization/compact/SchemaService.ts (hunk 10 above)
Comment (harunalpak) [reply src/serialization/compact/SchemaService.ts:104]: I agree with you. I forgot to remove catch block. I only use for testing. Thanks for attention.

I removed catch block.

File: src/serialization/compact/SchemaService.ts (hunk 14 above)
Comment (harunalpak) [reply src/serialization/compact/SchemaService.ts:103]: Fixed

File: src/serialization/compact/SchemaService.ts (hunk 11 above)
Comment (harunalpak) [reply src/serialization/compact/SchemaService.ts:138]: Fixed

File: src/serialization/compact/SchemaService.ts (hunk 13 above)
Comment (harunalpak) [reply src/serialization/compact/SchemaService.ts:141]: Fixed

File: src/serialization/compact/SchemaService.ts (hunk 12 above)
Comment (harunalpak) [reply src/serialization/compact/SchemaService.ts:154]: Fixed

**Comment by harunalpak** [comment]
> The coverage drop is 0.11%, which is not acceptable

I added a new test. And code coverage increased now.

File: test/integration/backward_compatible/serial/CompactRestartTest.js [hunk 15]
```diff
@@ -98,4 +101,28 @@ describe('CompactRestartTest', function() {
         const values = await map.valuesWithPredicate(Predicates.sql('INT32 == 42'));
         values.size().should.be.eq(1);
     });
+
+    it('should throw IllegalStateError exception, because of schema could not be replicated in the cluster', async function() {
```
Comment (srknzl) [inline test/integration/backward_compatible/serial/CompactRestartTest.js:105]: I think the test is ok, though we should put it in another file. `test/integration/backward_compatible/parallel/serialization/compact/CompactTest.js` should be better. Don't forget to skip the test if the client version is older than 5.2.0 because we added this behavior in 5.2.0.

**Review by srknzl — CHANGES_REQUESTED** [review]
One last thing:

**Review by srknzl — APPROVED** [review]
